src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +28,74 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
-        LOGGER.fine("Authentication attempted from "+username+" with "+key);
-        User u = User.get(username, false);
-        if (u==null) {
-            LOGGER.fine("No such user exists: "+username);
+        User user = this.retrieveOnlyKeyValidatedUser(username, key);
+
+        if (user == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
-        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
-        if (sshKey==null) {
-            LOGGER.fine("No SSH key registered for user: "+username);
+        Authentication auth = this.verifyUserUsingSecurityRealm(user);
+        if (auth == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
+        UserDetails userDetails = new SSHUserDetails(username, auth);
+        SecurityListener.fireAuthenticated(userDetails);
+        return true;
+    }
+
+    private @CheckForNull User retrieveOnlyKeyValidatedUser(String username, PublicKey key) {
+        LOGGER.log(Level.FINE, "Authentication attempted from {0} with {1}", new Object[]{ username, key });
+        User u = User.get(username, false, Collections.emptyMap());
+        if (u == null) {
+            LOGGER.log(Level.FINE, "No such user exists: {0}", new Object[]{ username });
+            return null;
+        }
+
+        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
+        if (sshKey == null) {
+            LOGGER.log(Level.FINE, "No SSH key registered for user: {0}", new Object[]{ username });
+            return null;
+        }
+
         String signature = signatureWriter.asString(key);
         if (!sshKey.isAuthorizedKey(signature)) {
-            LOGGER.fine("Key signature didn't match for the user: "+username+" : " + signature);
-            return false;
+            LOGGER.log(Level.FINE,"Key signature didn't match for the user: {0} : {1}", new Object[]{ username, signature });
```

Comment:
Careful with `'` in message formats. Use `didn’t` or just spell out `did not`.

Reply:
+1, better to fix it

Reply:
Can you please elaborate the reasons for that ?
I will also correct my other PR to follow that rule.

Reply:
See https://stackoverflow.com/questions/22670627/java-logger-apostrophe-issues-with-tokens

Reply:
Perfect, thank you 👍